### ISSUE TITLE
csc.exe crashing due to uncaught exception

### ISSUE THREAD

**Issue by murrple-1** [dotnet/roslyn#11992]
**Version Used**:
Microsoft Visual C# Compiler version 1.2.0.60317 (csc.exe)

**Steps to Reproduce**:

I'm not really certain, unfortunately. I have a relatively large codebase I just refactored and upon fixing the last compile error found by Intellisense (in Visual Studio 14.0.25123.00), the build step is now failing with this error.

**Expected Behavior**:
A regular build step (returning syntax errors, exit code 0, the usual)

**Actual Behavior**:
When csc.exe is run as a build step, all VS is reporting is `"csc.exe" exited with code -2146232797.`, which is probably not a useful exit code.

However, when running the command in the VS debugger, I am able to get this stack trace:

    Unhandled Exception: System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.BoundBadStatement' to type 'Microsoft.CodeAnalysis.CSharp.BoundStatementList'.
    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
    at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass20_0.<CompileNamedTypeAsTask>b__0()

If it's of any use, I tried reading the relevant source code, and the exception is probably being raised [at this line](https://github.com/dotnet/roslyn/blob/586f37ae8ee27162ea334a22d6b431cdd5e05a14/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs#L1045)

**Comment by murrple-1** [comment]
I have a little more info on this.

I was able to solve the error by commenting out pieces of code until I found the culprit, and I was able to fix it. I'll add it here in case it helps anyone in the future. I attempted to copy the relevant code into a reproducible minimal case, but was unsuccessful, so there's more going on than I am able to capture. However, I will note what I had to change here to workaround the crash.
```
new FieldMetadata<MetaJournal>()
{
    FieldName = "Title",
    ReturnDelegate = (mj, s) => { return (mj.Schema as dynamic).Title; }
},
```

was changed to 
```
new FieldMetadata<MetaJournal>()
{
    FieldName = "Title",
    ReturnDelegate = (mj, s) => { return mj.Schema.Title; }
},
```

Of note, `mj.Schema` returns a [`Newtonsoft.Json.Schema.JSchema`](http://www.newtonsoft.com/jsonschema/help/html/T_Newtonsoft_Json_Schema_JSchema.htm), which doesn't implement `DynamicObject`. `ReturnDelegate` is of type `Func<T, `[`MicroLite.IReadOnlySession`](https://github.com/TrevorPilley/MicroLite/blob/master/MicroLite/IReadOnlySession.cs)`, object>`.
